chainstate/tx-verifier/src/transaction_verifier/reward_distribution.rs

```diff
@@ -0,0 +1,575 @@
+
+                Ok(distribute_remainder_undo
+                    .unwrap_or_default()
+                    .into_iter()
+                    .chain(delegation_undos.into_iter())
+                    .collect())
+            } else {
+                // if total balance of all delegations is 0 then give the reward to the pool's owner
+                let increase_pool_balance_undo = accounting_adapter
+                    .operations(TransactionSource::Chain(block_id))
+                    .increase_pool_owner_balance(pool_id, total_delegations_reward)?;
+                Ok(vec![increase_pool_balance_undo])
+            }
+        })
+        .transpose()
+}
+
+fn calculate_rewards_per_delegation(
+    delegation_shares: &BTreeMap<DelegationId, Amount>,
+    block_id: Id<Block>,
+    total_delegations_balance: Amount,
+    total_delegations_reward: Amount,
+) -> Result<Vec<(DelegationId, u128)>, ConnectTransactionError> {
+    let total_delegations_balance = Uint256::from_amount(total_delegations_balance);
+    let total_delegations_reward = Uint256::from_amount(total_delegations_reward);
+    delegation_shares
+        .iter()
+        .map(
+            |(delegation_id, balance)| -> Result<_, ConnectTransactionError> {
+                let balance = Uint256::from_amount(*balance);
+                let reward = (total_delegations_reward * balance) / total_delegations_balance;
+                let reward: u128 = reward.try_into().map_err(|_| {
+                    ConnectTransactionError::DelegationRewardOverflow(*delegation_id, block_id)
+                })?;
+                Ok((*delegation_id, reward))
+            },
+        )
+        .collect::<Result<Vec<_>, _>>()
+}
+
+/// Due to integer arithmetics there can be a small remainder after all the delegations distributed.
+/// This remainder goes to the pool's owner
+fn distribute_delegations_reward_remainder<P: PoSAccountingView>(
+    accounting_adapter: &mut PoSAccountingDeltaAdapter<P>,
+    block_id: Id<Block>,
+    pool_id: PoolId,
+    total_delegations_reward: Amount,
+    rewards_per_delegation: &[(DelegationId, u128)],
+) -> Result<Option<Vec<PoSAccountingUndo>>, ConnectTransactionError> {
+    let total_delegations_reward_distributed = rewards_per_delegation
+        .iter()
+        .map(|(_, v)| Amount::from_atoms(*v))
+        .try_fold(Amount::ZERO, |acc, v| {
+            (acc + v).ok_or(ConnectTransactionError::DelegationsRewardSumFailed(
+                block_id, pool_id,
+            ))
+        })?;
+
+    let delegations_reward_remainder =
+        (total_delegations_reward - total_delegations_reward_distributed).ok_or(
+            ConnectTransactionError::DistributedDelegationsRewardExceedTotal(
+                pool_id,
+                block_id,
+                total_delegations_reward_distributed,
+                total_delegations_reward,
+            ),
+        )?;
+
+    if delegations_reward_remainder != Amount::ZERO {
```

Comment:
Why not just an empty vector? The sum of that is zero as well.

Reply:
It would just perform useless operation and produce useless undo. Maybe we should ignore 0 cases inside `pos_accounting`

Reply:
Then please make it with a subset/superset relationship. Not different branches.

Reply:
Do you mean like allow actually calling `increase_pool_pledge_amount(0)` with no-op as an undo? In other work hide this behavior inside `pos_accounting`?

Reply:
OK. So we have no Option anymore. All good.